adapter: simplify subscribe cancelation

benesch opened the pull request:
This commit implements the simplification to subscribe lifecycle identified by @teskje in [0]. Highlights:

  * The adapter now immediately removes its state about a subscribe when it aborts a subscribe, thus eliminating the need for `ActiveSubscription::dropping`.

      * The logic for dropping subscribes is no longer semi-circular. The `drop_compute_sinks` method triggers a call to `drop_collections`, which triggers `SubscribeResponse::DroppedAt`, which does *not* cause another call to `drop_compute_sinks`. The only way that adapter and controller state gets cleaned up now is via a call to `drop_compute_sinks`.

  * The separate cancellation channel between the adapter and the protocol layers is no more. Cancellation is now *only* communicated via the peek and subscribe response protocol. This makes the protocol much easier to follow, with seemingly no loss of functionality or ability to send cancellation requests eagerly.

[0]: https://github.com/MaterializeInc/materialize/issues/21081#issuecomment-1903644620

Fix #16121.
Fix #21081.


### Motivation

   * This PR fixes outstanding bugs.
   * This PR refactors existing code.


### Tips for reviewer

Suppressing whitespace goes a long way.

@teskje please see line comments within; there's one piece I'm very not sure about.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [x] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  <!-- Reference the design in the description. -->
- [x] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [x] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  <!-- Ask in #team-cloud on Slack if you need help preparing the cloud PR. -->
- [x] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - n/a

benesch commented on src/adapter/src/coord/ddl.rs:
```diff
@@ -751,23 +736,10 @@ impl Coordinator {
+            self.remove_active_subscribe(sink.global_id, reason).await;
 
             if !self
                 .controller
```
@teskje there's now a code path where the compute read policy is not removed. If a `SUBSCRIBE ... UP TO` completes, then it will be removed from the adapter state when the `SubscribeResponse::DroppedAt` is processed... but `drop_compute_sinks` will thus never get called on that subscribe. That seems like a problem, and it's related to the next comment...

benesch replied:
https://github.com/MaterializeInc/materialize/pull/25179#discussion_r1487243233

benesch commented on src/adapter/src/coord/ddl.rs:
```diff
@@ -786,11 +758,6 @@ impl Coordinator {
-            }
         }
         let mut compute = self.controller.active_compute();
         for (cluster_id, ids) in by_cluster {
```
...in the same situation as described in the last comment, we'll never call `drop_collections` on said subscribe. Is that a problem, or is the controller smart enough to clean up its internal state when it observes a `SubscribeResponse::Dropped`?

benesch replied:
https://github.com/MaterializeInc/materialize/pull/25179#discussion_r1487243233

benesch commented on src/adapter/src/coord/message_handler.rs:
```diff
@@ -346,18 +344,15 @@ impl Coordinator {
-                    let remove = active_subscribe.process_response(response);
-                    if remove {
-                        let csid = ComputeSinkId {
-                            cluster_id: active_subscribe.cluster_id,
-                            global_id: sink_id,
-                        };
-                        self.drop_compute_sinks([csid]);
-                        self.remove_active_subscribe(sink_id).await;
+                    let finished = active_subscribe.process_response(response);
+                    if finished {
+                        self.remove_active_subscribe(sink_id, SubscribeRemovalReason::Finished)
+                            .await;
```
@teskje if it _is_ important to call `drop_collections` on all subscribes, I wonder if it might be better to:

  * Never clean up coordinator state here, in response to a `SubscribeResponse` from the controller.
  * Instead, rely on the coordinator to call `drop_compute_sinks` on every subscribe exactly once, either because of a cancellation request, or because one of the dependencies was dropped, or because the subscribe completed successfully and the transaction containing it was committed.

benesch replied:
I talked myself into this being compelling enough of a take that I've updated the implementation accordingly.

teskje replied:
You figured this out yourself already, but just to confirm: It is important that adapter calls `drop_collections` for all subscribes, even ones that have advanced to the empty frontier. The controller keeps some collection state around until the client has indicated that it doesn't need the collection anymore.

Your approach sounds exactly right to me!

benesch commented:
Test failures here all look expected. Won't have time to clean them up tonight. Would be delighted if someone reviewing has spare cycles to help get the tests passing.

moulimukherjee commented on src/adapter/src/coord/message_handler.rs:
```diff
@@ -346,19 +343,12 @@ impl Coordinator {
-                        self.drop_compute_sinks([csid]);
-                        self.remove_active_subscribe(sink_id).await;
-                    }
+                    active_subscribe.process_response(response);
```
If the cluster is dropped, would it leak state?

I see we are doing a `drop_compute_sinks(...).await` before dropping the cluster. Could the cluster die before a `DroppedAt` message is sent back?

teskje replied:
> This seemed the sanest way to handle all this to me. The adapter is guaranteed to call `create_dataflow` and `drop_collection` for each subscribe exactly once, and doesn't need to do any complicated bookkeeping. But maybe there are some defects that I'm missing.

Sounds good to me! The only "defect" that I'm seeing is that the controller might keep the collection state around longer than necessary for a finished subscribe if the transaction takes time to complete. But I think that's something we can live with and I wouldn't want to make the code more complicated to do something about it.

> I see we are doing a `drop_compute_sinks(...).await` before dropping the cluster. Could the cluster die before a `DroppedAt` message is sent back which would have cleared the active subscribe state?

What active subscribe state are you referring to? Adapter's state is removed as part of `drop_compute_sinks` (and adapter doesn't receive `DroppedAt` responses anyway anymore). The only part that relies on the `DroppedAt` responses for state cleanup is the compute networking stack (specifically `PartitionedComputeState`), but that is per-cluster state which gets dropped entirely when we drop the cluster, so no need to still clean anything up.

moulimukherjee replied:
@teskje Ah right, `drop_compute_sinks` now clear the adapter's active_subscribes state. Sorry, overlooked it 🤦‍♀️!

moulimukherjee replied:
Maybe dumb q, how are we removing entry from adapter's `active_subscribes` after we receive say the final response? Is that via the `clear_connection` method?

benesch replied:
> Maybe dumb q, how are we removing entry from adapter's `active_subscribes` after we receive say the final response? Is that via the `clear_connection` method?

Not a dumb question. That's exactly right! It's via `clear_connection`. When the subscribe completes successfully, the client will either commit the transaction, thus resulting in a call `clear_connection`, or end the session, again resulting in a call to `clear_connect`ion.

teskje commented:
I've taken the freedom of pushing another commit that simplifies the controller API by not emitting responses for dropped subscribes anymore. That makes it so that adapter doesn't have to worry about late responses or `DroppedAt` responses. Doesn't have a big effect on the current adapter implementation it seems, but it still makes things easier to reason about, I hope!

jkosh44 reviewed (approved):
Adapter code LGTM

moulimukherjee commented on src/environmentd/tests/sql.rs:
```diff
@@ -2598,7 +2598,7 @@ fn test_subscribe_on_dropped_source() {
         }
 
-        query_thread.join().unwrap();
+        let _ = query_thread.join();
```
There's no notice anymore, instead the response has the error that `the subscribe was dropped...`

benesch replied:
Right, but removing this `unwrap` means the test succeeds when it shouldn't! As written it is attempting to assert the presence of a notice that never comes, but the test still passes. I've pushed up a new revision that restores the `unwrap` and adjusts `assert_subscribe_notice` to not check for the notice that never comes.

moulimukherjee replied:
Oh I see, makes sense.

benesch commented:
> I've taken the freedom of pushing another commit that simplifies the controller API by not emitting responses for dropped subscribes anymore. That makes it so that adapter doesn't have to worry about late responses or `DroppedAt` responses. Doesn't have a big effect on the current adapter implementation it seems, but it still makes things easier to reason about, I hope!

I was so hoping you might do that. Thank you!

benesch reviewed (commented):
Just pushed up a revision that I think will pass all tests. Thanks for all the help, y'all!

teskje reviewed (approved):
Compute changes lgtm (but I also wrote them :D). I don't understand the adapter changes in detail, but on a high level they seem very sensible!

def- commented:
The test failure is fixed on main, so with a rebase this should merge. The parallel-workload requires a new error to be ignored, I'll add that now.

benesch commented:
> The test failure is fixed on main, so with a rebase this should merge. The parallel-workload requires a new error to be ignored, I'll add that now.

Gorgeous, thank you! I think your last push didn't have the changes from last night, though. I squashed the parallel-workloads fix into what I had locally, rebased, and force pushed. I think that should bring all the fixes together. 🤞🏽

nrainer-materialize commented:
I had triggered a [nightly build:](https://buildkite.com/materialize/nightlies/builds/6430)
[Parallel Workload (100 threads)](https://buildkite.com/materialize/nightlies/builds/6430#018da7a4-9d3d-410e-8779-7b4858fc2b99) fails and this does not seem like a flake. @def-, is this a test issue and being fixed by adding an ignore?

def- commented:
@nrainer-materialize Should be fixed already: https://github.com/MaterializeInc/materialize/pull/25179/files#diff-c59bf417ba5c2310e001f981beb799c9cf27603e1aca332692390b278800c264
